### PR TITLE
Icmslst 665 Refactor FIR urls and views

### PR DESCRIPTION
**Pull request by MattHolmes123**
Changes:
  - Refactor manage-firs urls and views.
  - Refactor list-firs urls and views.
  - Refactor add-fir urls and views.
  - Refactor edit-fir urls and views.
  - Refactor respond-fir urls and views.
  - Refactor archive-fir urls and views and rename to delete-fir.
  - Refactor withdraw-fir urls and views.
  - Refactor close-fir urls and views.
  - Refactor archive-fir-file and rename to delete-fir-file.
  - Add FIR functionality to export applications.

### CODE REVIEW

File: web/domains/case/views.py [hunk 1]
```diff
@@ -497,18 +498,36 @@ def _close_update_requests(request, application_pk, update_request_pk, model_cla
...
 
 
-def _manage_firs(request, application_pk, model_class, url_namespace, **extra_context):
+@login_required
+@permission_required("web.reference_data_access", raise_exception=True)
+def manage_firs(request: HttpRequest, *, application_pk: int, case_type: str) -> HttpResponse:
+    model_class = _get_class_imp_or_exp_or_access(case_type)
+
     with transaction.atomic():
         application = get_object_or_404(model_class.objects.select_for_update(), pk=application_pk)
```
Comment (oskusalerma) [inline web/domains/case/views.py:507]: do you want to add the type hints to `application` in all of these functions you've touched?

File: web/domains/case/views.py [hunk 2]
```diff
@@ -541,19 +574,34 @@ def _add_fir(request, application_pk, model_class, url_namespace):
...
+def edit_fir(request, *, application_pk: int, fir_pk: int, case_type: str) -> HttpResponse:
     with transaction.atomic():
+        model_class = _get_class_imp_or_exp_or_access(case_type)
         application = get_object_or_404(model_class.objects.select_for_update(), pk=application_pk)
+
+        if case_type == "access":
+            contacts = [application.submitted_by]
+
+        elif case_type == "import":
+            application = get_object_or_404(model_class, pk=application_pk)
```
Comment (oskusalerma) [inline web/domains/case/views.py:591]: is this not the same as line 585...?

File: web/domains/case/views.py [hunk 3]
```diff
@@ -541,19 +574,34 @@ def _add_fir(request, application_pk, model_class, url_namespace):
...
+            contacts = get_users_with_perms(
+                application.importer, only_with_perms_in=["is_contact_of_importer"]
+            ).filter(user_permissions__codename="importer_access")
+
+        elif case_type == "export":
+            application = get_object_or_404(model_class, pk=application_pk)
```
Comment (oskusalerma) [inline web/domains/case/views.py:597]: is this not the same as line 585...?

File: web/templates/partial/case/case-fir-box-user.html [hunk 4]
```diff
@@ -1,10 +1,11 @@
+{# TODO: ICMSLST-670 remove url_namespace from file #}
 {% extends "partial/case/fir.html" %}
 
 {% block fir_action %}
   {% if fir.status == "OPEN" %}
     <a href="{{ icms_url(
-                  url_namespace + ':respond-fir',
-                  kwargs={'application_pk': process.pk, 'fir_pk': fir.pk }) }}"
+                  'case:respond-fir',
+                  kwargs={'application_pk': process.pk, 'fir_pk': fir.pk, 'case_type': url_namespace }) }}"
```
Comment (oskusalerma) [inline web/templates/partial/case/case-fir-box-user.html:8]: why are you not using case_type here and in all other fir template files already?

**Review by oskusalerma — CHANGES_REQUESTED** [review]

File: web/templates/partial/case/case-fir-box-user.html (hunk 4 above)
Comment (MattHolmes123) [reply web/templates/partial/case/case-fir-box-user.html:8]: Only reason is I didn't want to change anything as part of the refactor.
I couldn't guarantee that case_type was going to be available (without further checking / testing).

File: web/domains/case/views.py (hunk 3 above)
Comment (MattHolmes123) [reply web/domains/case/views.py:597]: Yep

File: web/domains/case/views.py (hunk 2 above)
Comment (MattHolmes123) [reply web/domains/case/views.py:591]: Yep

**Review by oskusalerma — APPROVED** [review]

**Pull request merged**